File: Templates/MAX Segment Tree.java
class SegmentTreeTemplate{

    // Array Implementation of Segment Tree ------ Max Queries ONLY ------
    private int[] ST;
    private int n;

    public SegmentTreeTemplate(int[] arr){
		n = arr.length;
		ST = new int[n << 1];
        System.arraycopy(arr, 0, ST, n, n);
        for(int i = n - 1; i > 0; i--){
            ST[i] = Math.max(ST[i << 1], ST[(i << 1) + 1]);
        }
	}

    public boolean update(int p, int newVal){
        p += n;
        ST[p] = newVal;
        while (p > 1){
            p >>= 1;
            newVal = Math.max(ST[p << 1], ST[(p << 1) + 1]);
            if(newVal != ST[p]){
                ST[p] = newVal;
            }
            else{
                return true;
            }
        }
        return true;
    }

    public int getMax(int l, int r){
        l += n;
        r += n;
        int max = Integer.MIN_VALUE;
        while (l < r){
            if((l & 1) == 1){
                max = Math.max(max, ST[l]);
                l++;
            }
            if ((r & 1) == 1){
                max = Math.max(max, ST[r]);
                r--;
            }
            r >>= 1;
            l >>= 1;
        }
        return max;
    }
}

File: Templates/MAX Segment Tree.py
# Segment Tree for MAX Queries
class SegmentTreeTemplate:

    def __init__(self, arr: list):
        self.n = len(arr)
        self.ST = [0] * (self.n << 1)
        self.ST[self.n:] = arr[0:self.n]
        for i in range(self.n - 1, -1, -1):
            self.ST[i] = max(self.ST[i << 1], self.ST[(i << 1) + 1])

    def update(self, p: int, newVal: int) -> bool:
        p += self.n
        self.ST[p] = newVal
        while p > 1:
            p >>= 1
            newVal = max(self.ST[p << 1], self.ST[(p << 1) + 1])
            if newVal != self.ST[p]:
                self.ST[p] = newVal
            else:
                return True
        return True

    def getMax(self, l: int, r: int) -> int:
        l += self.n
        r += self.n
        maxx = float('-inf')
        while l < r:
            if (l & 1) == 1:
                maxx = max(maxx, self.ST[l])
                l += 1
            if (r & 1) == 1:
                r -= 1
                maxx = max(maxx, self.ST[r])
            r >>= 1
            l >>= 1
        return maxx